Sources/Networking/Core/Upload/UploadAPIManager.swift

```diff
@@ -0,0 +1,286 @@
+            sessionUploadTask.resume()
+            return uploadTask
+        } catch {
+            throw await errorProcessors.process(error, for: request)
+        }
+    }
+
+    /// Returns any stored upload task and updates its internal URLSessionUploadTask, or creates a new one.
+    func existingUploadTaskOrNew(
+        for sessionUploadTask: URLSessionUploadTask,
+        request: EndpointRequest,
+        uploadable: Uploadable
+    ) async -> UploadTask {
+        guard var existingUploadTask = await uploadTasks.getValue(for: request.id) else {
+            return UploadTask(
+                sessionUploadTask: sessionUploadTask,
+                endpointRequest: request,
+                uploadable: uploadable
+            )
+        }
+        existingUploadTask.task = sessionUploadTask
+        return existingUploadTask
+    }
+
+    func handleUploadTaskCompletion(
+        urlRequest: URLRequest,
+        endpointRequest: EndpointRequest,
+        retryConfiguration: RetryConfiguration?,
+        data: Data?,
+        response: URLResponse?,
+        error: Error?
+    ) {
+        Task {
+            guard let uploadTask = await uploadTasks.getValue(for: endpointRequest.id) else {
+                return
+            }
+
+            var state = UploadTask.State(task: uploadTask.task)
+            if let data, let response {
+                state.response = try await responseProcessors.process(
+                    (data, response),
+                    with: urlRequest,
+                    for: endpointRequest
+                )
+
+                try await uploadTask.complete(with: state)
+
+                // Cleanup on successful task completion
+                await uploadTask.resetRetryCounter()
+                await uploadTasks.set(value: nil, for: endpointRequest.id)
+            } else if let error {
+                do {
+                    // URLError.Code.cancelled is thrown if the URLSessionTask is cancelled.
+                    // Consider this action intentional, thus the request won't be retried.
+                    guard !state.cancelled else {
+                        throw error
+                    }
+
+                    try await uploadTask.sleepIfRetry(
+                        for: error,
+                        retryConfiguration: retryConfiguration
+                    )
+
+                    try await self.uploadRequest(
+                        uploadTask.uploadable,
+                        request: uploadTask.endpointRequest,
+                        retryConfiguration: retryConfiguration
+                    )
+                } catch {
+                    state.error = await errorProcessors.process(
+                        error,
+                        for: uploadTask.endpointRequest
+                    )
+
+                    // No cleanup in case the task will be retried.
+                    try await uploadTask.complete(with: state)
```

Comment:
If the user never retries by using the retry method, will the task ever get removed from the uploadTasks if it fails?

Reply:
Currently only when the manager is deinitialized. What do you think would be the best to handle this?

Reply:
There are only two ways I guess, either we have retry functionality as is but keep hoarding failed requests (which in case of singletons is not great) or we don't have the retry function and just remove it from uploadTasks after it fails. :(

Reply:
When implementing, I had some doubts regarding the expected retry behavior for uploads. Is it even common practice to automatically retry failed uploads? For instance, if an upload request fails in the middle of the process and is retried without providing any feedback to the user, it might appear strange that the progress suddenly jumps from 80% back to 10%. 

Considering this case, it might be better to leave the retry behavior as the dev's responsibility. From the library's perspective, any upload request that errors out would be canceled (as you mentioned). This approach allows clients/devs to have more control over the retry logic according to their specific requirements.

Reply:
Agree with @tonyskansf , would be very helpful to have retry functionality in place, at the same time it's true we don't know use case behaviour. 
When uploading in chunks retry should continue with last successful chunk +1 etc. 
For sake of simplicity I would go with option that upload doesn't have retry at all so it's dev responsibility eventually

Reply:
> When uploading in chunks retry should continue with last successful chunk +1 etc.

I believe this depends on the server's implementation too — if it supports partial (resumable) upload.